communication/src/communication.cpp

```
@@ -0,0 +1,169 @@
+#include "communication.h"
+#include <future>
+
+Communication* Communication::instance = nullptr;
+
+// Constructor
+Communication::Communication(uint32_t id, void (*passDataCallback)(void *)) : 
+    client(std::bind(&Communication::receivePacket, this, std::placeholders::_1))
+{
+    setId(id);
+    setPassDataCallback(passDataCallback);
+
+    instance = this;
+
+    auto signalResult = signal(SIGINT, Communication::signalHandler);
+    if (signalResult == SIG_ERR)
+        throw std::runtime_error("Failed to set signal handler for SIGINT");
```

Should return an error code?